Monthly stepsize check

philipjdavis opened the pull request:
Hi @arjclark

Can you kindly review these changes, which addresses #10?

## Summary of changes
These changes allow the script to check that the data step sizes in monthly files are as expected, based on the model initialisation date and run length. This is because the step size will change according to the months being covered by a given model run. In addition, supporting documentation has been updated, and the CONTRIBUTOR.md file has been updated with my name.

Specific are below:

_CONFIG_FORMAT.md_
* Updated the section on required_intervals for the case of monthly files,
* Added description on the allowed_dimensions structure which was previously missing,

_CONTRIBUTING.md_
* File has been signed

_README.md_
* Updated to include new optional arguments initdate and runlength, needed to calculate the expected monthly data intervals,

_example.json_
* Updated to include an example entry for "allowed_dimensions"

_ncdfchecker.py_
* https://github.com/philipjdavis/ncdfchecker/blob/96b5afa4f1afec81429c7197e196435be59e9af0/ncdfchecker.py#L172 : flake8 found a bug in this variable name, which is now fixed,
* https://github.com/philipjdavis/ncdfchecker/blob/96b5afa4f1afec81429c7197e196435be59e9af0/ncdfchecker.py#L196-L234 : Given the model initialisation date and run length, `get_expected_monthly_stepsize` calculates
  * the complete months that are covered by the model run,
  * the mid-points of those months (which are the dates specified by the leadtime in the monthly file), calculated by `get_month_midpoints`
  * the interval between these values, and are returned in hours
* `check_stepsize` has been update to handle monthly files, which compares the array of expected intervals with those found in the file
* `simple_variable_checks` updated to handle step-checking for monthly files, if the required_intervals entry for a variable is set to "mon",
* Added new optional arguments initdate and runlength which respectively give the model's initialisation date and run length,

_test_ncdfchecker.py_
* Added unit tests for `check_stepsize`, `simple_variable_checks`, `load_constraints` for the case where a json file contains a monthly variable. The latter is created by substituting the 6-hourly test field with a monthly one.
* Added unit tests for the `get_expected_monthly_stepsize` and `get_month_midpoints`.

# Checks
* All unit tests pass, and pycodestyle returns no errors **except** regarding the bare `exception` on https://github.com/philipjdavis/ncdfchecker/blob/96b5afa4f1afec81429c7197e196435be59e9af0/test_ncdfchecker.py#L346. I think this is ok because it is catching something unexpected.
* Tested in the GloSea suite and gs[fh]c_c3s_checks tasks pass.

philipjdavis committed:
Added functionality to check data step sizes for monthly files. Added corresponding unit tests.

philipjdavis committed:
Updated documentation: signed CONTRIBUTING.md, updated CONFIG_FORMAT to describe step-ize checks for monthly files and describe allowed_dimension, updated README.md to reflect new argument changes.

arjclark commented on README.md:
```diff
@@ -4,7 +4,7 @@ ncdfchecker.py is a config driven utility that will allow you to check that a
 provided in a json file.
 
-usage: ncdfchecker.py [-h] [--quiet] [--strict] input_file config_path
+usage: ncdfchecker.py [-h] [--quiet] [--strict] [--initdate INITDATE] [--runlength RUNLENGTH] input_file config_path
```
If you are providing command line overrides there needs to be an equivalent config entry - the tool is first and foremost config file driven.

Runlength is not suitable in a generic case - again, in the spirit of the tool rather than the specific application you're planning on using it for. min_date and max_date would be in line with naming conventions and structure elsewhere.

philipjdavis replied:
As discussed offline, the changes now only use the "forecast_reference_time" in the NetCDF meta-data to convert the leadtimes in the file to datetime objects so that we can grab the required period. I have therefore removed these extra arguments from the checker script.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
+    return midmonths
+
+
+def get_expected_monthly_stepsize(initdate, runlength):
```
Initdate should be taken from the first entry in the item being validated - so is startdate rather than GloSea model based nomenclature of init date.

philipjdavis replied:
As mentioned above the start date is taken from the meta-data "forecast_reference_time".

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
     """
-    Helper routine to check that the stepsize in some given data is equal to
-    a given value
+    Using the model initialisation date and run length, calculate the
```
No models here - this is purely data driven with no knowledge of source, be it model, obs or otherwise.

philipjdavis replied:
Here, as elsewhere, I have removed any need of relying on externally specifying the start date and model run length.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
+    """
+    SECS_IN_HR = 60 * 60
+
+    initdate = dateutil.parser.parse(initdate)
```
Use first date entry from the data provided for iteration.

philipjdavis replied:
The forecast_reference_time is now used.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
+    SECS_IN_HR = 60 * 60
+
+    initdate = dateutil.parser.parse(initdate)
+    enddate = initdate + datetime.timedelta(days=runlength)
```
This can be taken from the data provided based on how you're using it.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
+            for timedelta in expected_stepsize]
+
+
+def check_stepsize(data, stepsize, monthly=False):
```
Don't tie to specific period type - fixed_period="unset" and insert "monthly" to trigger off your specific checks. If block so hangs off "monthly", "unset", and else to not implemented error at minimum here.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
 
-def simple_variable_checks(product, constraints, strict=False, logger=None):
+
+def simple_variable_checks(product, constraints, strict=False, logger=None,
```
See comments elsewhere about initdate and runlength

arjclark commented on ncdfchecker.py:
```diff
@@ -280,7 +334,22 @@ def simple_variable_checks(product, constraints, strict=False, logger=None):
                                 step = constraints[variable][key][intervalkey]
-                                if not check_stepsize(arr, step):
+
+                                monthly = step == 'mon'
```
Based on your use of step == 'mon' there is no need for the setting of monthly.

arjclark commented on ncdfchecker.py:
```diff
@@ -280,7 +334,22 @@ def simple_variable_checks(product, constraints, strict=False, logger=None):
+                                monthly = step == 'mon'
+                                if monthly:
+                                    logger.info('This is a monthly file.')
+                                    if initdate is None or runlength is None:
```
This doesn't make sense for checking if dates are monthly, it only makes sense for a validity date range - again though, as mentioned elsewhere I'm not sure that's necessary here.

arjclark commented on ncdfchecker.py:
```diff
@@ -280,7 +334,22 @@ def simple_variable_checks(product, constraints, strict=False, logger=None):
+                                    step = get_expected_monthly_stepsize(
+                                        initdate, runlength)
+
+                                if not check_stepsize(arr, step, monthly):
```
"monthly" not necessary here, already captured in step.

arjclark commented on ncdfchecker.py:
```diff
@@ -376,6 +445,10 @@ def simple_variable_checks(product, constraints, strict=False, logger=None):
                         help="If an unknown variable is encountered or fill "
                         "values specified but not used then it should be "
                         "considered an error.")
+    parser.add_argument('--initdate', '-i',
```
I don't see the point in these in order to enforce monthly midpoint dates. If there is a requirement to enforce validity dates outside of the monthly midpoint then this makes sense, otherwise please remove (see comments elsewhere though about needing a config entry if you wish to preserve these)

arjclark replied:
This might actually already be possible through the range controls on the time coordinate anyway, so should probably feed into that as an override and insert entries solution.

arjclark commented on ncdfchecker.py:
```diff
@@ -189,16 +193,66 @@ def check_globals(product, constraints, skip=["short_name"], strict=False,
     steparr = data[1:len(data)] - data[0:len(data)-1]
-    return np.all(steparr == stepsize)
 
+    if monthly:
```
if step=='mon'

arjclark reviewed (changes requested):
Afraid am going to need to bounce this back your way for quite a few changes (see comments).

More fundamentally though I think you're trying to both overengineer things here and tailor to your specific use case in GloSea which is not how the tool is intended for use. A much simpler approach is open to you.

1. take your data, whatever that is
2. extract month entries from the dates (leave open for day and year implementation, raise notimplemented errors as needed)
3. generate midpoints from those (again leave open for day and year implementation, raise not implemented errors as needed). If, for convenience you want a start and end date you can compare the dates in the data to get the min and max values (which would also protect you against missing entries) - arguably entry 0 and -1 should always be the min and max in a well formed coordinate system so you could just use that.
4. make sure entries in the data compare correctly with expected list

You also need to be explicit in your documentation that this implementation is using mid month points.

arjclark commented:
You may even be able to get away with a slightly simpler check of get the month number, make sure they're consecutive and then validate that each of the midpoints are correct dates.

philipjdavis committed:
Simplified and generalised the interval checking by grabbing the required period from the leadtimes

philipjdavis committed:
Some further simplifying of the get_period_stepsize function, and added additional comments. Updated README.md

philipjdavis committed:
Added unit tests for yearly files; create_example_files() now removed outside of classes for multiple use with other unit tests; additional tidying up in ncdfchecker.py

philipjdavis committed:
Minor tweaks to test_ncdfchecker.py

philipjdavis committed:
Updated CONFIG_FORMAT.md to describe yearly file checks; removed debug print statement

philipjdavis commented:
Hi @arjclark ,

As discussed offline, I have made the following simplifications based on your suggestion that we check the interval between month numbers (or years for yearly files):

[CONFIG_FORMAT.md](https://github.com/metosfc/ncdfchecker/compare/master...philipjdavis:monthly-stepsize-check#diff-b17b26932bd6b2a72ff6aeb859a6776c2185babd06380db0b36d1f23587e0768R113):
Updated the documentation to mention that checks for yearly files can also be done

***ncdfchecker.py***
* Added [get_period_stepsize()](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-0f54bb1d1d419aea51519a9b279677ea3d057f08a1b755d6a5123fe3587154dfR196), which uses the forecast_reference_time in the NetCDF meta-data to convert the lead times (in hours since the forecast start date) into datetime objects. The required time period is then extracted and the interval calculated,

* [check_stepsize()](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-0f54bb1d1d419aea51519a9b279677ea3d057f08a1b755d6a5123fe3587154dfR226) has been updated to handle temporal step-size checking, as well as purely between neighbouring data points needed for, e.g. lat-lon checks,

* [simple_variable_checks()](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-0f54bb1d1d419aea51519a9b279677ea3d057f08a1b755d6a5123fe3587154dfR242) has been updated to handle monthly or yearly checks. If "month" or "year" is specified in the json file, then we check that the step-size is 1. As discussed, this can be easily generalised to allow checks for arbitrary time periods and step-sizes. Will raise a separate issue. I have also removed the ```initdate``` and ```runlength``` arguments, 

***test_ncdfchecker.py***
* Added [forecast_reference_time](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R122) to sample json file,
* On top of the hourly and monthly product validation unit tests, I have also added similar tests for the [yearly](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R437) files. In order to easily generate example files for these, I have taken [create_example_files()](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R132) outside of the ```TestProductValidator``` class. I have also added arguments to specify time coordinate values and frequency. 
* Is easier to just update the ```testfield``` entries with the required values within the unit tests [e.g. here](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R367), so I have removed the separate [monthly_field](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3L130) entry,
* Updated [unmet_constraints](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R312) in ```TestProductValidator``` to include ```forecast_reference_time```,
* Removed unit tests for defunct ```get_month_midpoints``` and ```get_expected_monthly_stepsize```, and instead replaced with unit tests for [get_period_stepize](https://github.com/metosfc/ncdfchecker/compare/ef77b...1c6f#diff-914d75e02354e7aa2291191bd96fcf105f2cc47d4679056e8802939e9ca0cac3R514). This also includes a unit test for checking yearly step-sizes.

Cheers,

Phil

arjclark reviewed (approved):
Hi Phil - this looks much cleaner thanks.
